File: sponge-kotlin/src/test/kotlin/examples/FiltersBuilder.kt
/*
 * Sponge Knowledge base
 * Using filter builders
 */

package org.openksavi.sponge.kotlin.examples

import org.openksavi.sponge.event.Event
import org.openksavi.sponge.kotlin.KFilter
import org.openksavi.sponge.kotlin.KKnowledgeBase
import org.openksavi.sponge.kotlin.KTrigger
import java.util.Collections
import java.util.concurrent.atomic.AtomicInteger
import org.openksavi.sponge.kotlin.KFilterBuilder

class FiltersBuilder : KKnowledgeBase() {

    override fun onInit() {
        // Variables for assertions only
        var eventCounter = Collections.synchronizedMap(HashMap<String, AtomicInteger>())
        eventCounter.put("blue", AtomicInteger(0))
        eventCounter.put("red", AtomicInteger(0))
        sponge.setVariable("eventCounter", eventCounter)
    }

    class ColorTrigger : KTrigger() {
        override fun onConfigure() {
            withEvent("e1")
        }

        override fun onRun(event: Event) {
            logger.debug("Received event {}", event)
            sponge.getVariable<Map<String, AtomicInteger>>("eventCounter").get(event.get<String>("color"))!!.incrementAndGet()
        }
    }

    override fun onLoad() {
        sponge.enable(KFilterBuilder("ColorFilter").withEvent("e1").withOnAccept({ filter, event ->
            sponge.logger.debug("Received event {}", event)
            val color: String? = event.get("color", null)
            if (color == null || color != "blue") {
                sponge.logger.debug("rejected")
                false
            } else {
                sponge.logger.debug("accepted")
                true
            }
        }))
    }

    override fun onStartup() {
        sponge.event("e1").send()
        sponge.event("e1").set("color", "red").send()
        sponge.event("e1").set("color", "blue").send()
    }
}

File: sponge-nashorn/examples/script/js/filters_builder.js
/**
 * Sponge Knowledge Base
 * Using filter builders
 */

var AtomicInteger = java.util.concurrent.atomic.AtomicInteger

function onInit() {
    // Variables for assertions only
    var eventCounter = java.util.Collections.synchronizedMap(new java.util.HashMap())
    eventCounter.put("blue", new AtomicInteger(0))
    eventCounter.put("red", new AtomicInteger(0))
    sponge.setVariable("eventCounter", eventCounter)
}

var ColorTrigger = Java.extend(Trigger, {
    onConfigure: function(self) {
        self.withEvent("e1")
    },
    onRun: function(self, event) {
        self.logger.debug("Received event {}", event)
        sponge.getVariable("eventCounter").get(event.get("color")).incrementAndGet()
    }
});

function onLoad() {
    sponge.enable(new FilterBuilder("ColorFilter").withEvent("e1").withOnAccept(function (filter, event) {
        sponge.logger.debug("Received event {}", event)
        var color = event.get("color", null)
        if (color == null || color != "blue") {
            sponge.logger.debug("rejected")
            return false
        } else {
            sponge.logger.debug("accepted")
            return true
        }
    }))
}

function onStartup() {
    sponge.event("e1").send()
    sponge.event("e1").set("color", "red").send()
    sponge.event("e1").set("color", "blue").send()
}
